### PR TITLE
Remove ReferenceMap from top-level frontend passes

### PR DESCRIPTION
**Pull request by asl**
Finally, this removes ReferenceMap from top-level frontend passes:
 - RemoveAllUnusedDeclarations uses lightweight implementation of used set that was factored out from `ResolveReferences`
 - Added shortcut `CheckShadow` to execute `ResolveReferences` with shadow check enabled (and with "local" refmap)
 - Made refmap and internal implementation details of Evaluator, Inliner and SimplifyDefUse

### CODE REVIEW

File: frontends/p4/unusedDeclarations.cpp [hunk 1]
```diff
@@ -16,21 +16,50 @@ limitations under the License.
...
 RemoveUnusedDeclarations *RemoveUnusedPolicy::getRemoveUnusedDeclarationsPass(
-    const ReferenceMap *refMap, bool warn) const {
-    return new RemoveUnusedDeclarations(refMap, warn);
+    const UsedDeclSet &used, bool warn) const {
+    return new RemoveUnusedDeclarations(used, warn);
 }
 
 Visitor::profile_t RemoveUnusedDeclarations::init_apply(const IR::Node *node) {
-    LOG4("Reference map " << refMap);
+    LOG4("Used sed " << used);
```
Comment (asl) [inline frontends/p4/unusedDeclarations.cpp:29]:
```suggestion
    LOG4("Used set " << used);
```

File: frontends/p4/redundantParsers.h [hunk 2]
```diff
@@ -38,27 +39,26 @@ class FindRedundantParsers : public Inspector {
...
 };
 
 class RemoveRedundantParsers : public PassManager {
     std::set<const IR::P4Parser *> redundantParsers;
 
  public:
-    RemoveRedundantParsers(ReferenceMap *refMap, TypeMap *typeMap, const RemoveUnusedPolicy &policy)
-        : PassManager{new TypeChecking(refMap, typeMap, true),
+    RemoveRedundantParsers(TypeMap *typeMap, const RemoveUnusedPolicy &policy)
+        : PassManager{new TypeChecking(nullptr, typeMap, true),
```
Comment (fruffy) [inline frontends/p4/redundantParsers.h:58]: I missed this in the other PRs but passing in nullptr doesn't seem right to me. Any way to make the constructor safe and just use two parameters? I also ran into this with ConstantFolding, which now has an ambiguous constructor.

File: frontends/common/resolveReferences/resolveReferences.h [hunk 3]
```diff
@@ -170,6 +172,18 @@ class ResolveReferences : public Inspector, private ResolutionContext {
     void checkShadowing(const IR::INamespace *ns) const;
 };
 
+class CheckShadow : public PassManager {
```
Comment (vlstill) [inline frontends/common/resolveReferences/resolveReferences.h:175]: Maybe `CheckShadowing` would be a better name?

File: frontends/p4/evaluator/evaluator.h [hunk 4]
```diff
@@ -113,10 +114,12 @@ class Evaluator final : public Inspector, public IHasBlock {
 /// high-level constructs.
 class EvaluatorPass final : public PassManager, public IHasBlock {
     P4::Evaluator *evaluator;
+    ReferenceMap selfRefMap;
```
Comment (vlstill) [inline frontends/p4/evaluator/evaluator.h:117]: Out of curiosity, why not hold the local map in a `unique_ptr`? It seems like quite big struct to have it created in all cases and the access is always through pointer anyway.

File: frontends/p4/unusedDeclarations.h [hunk 5]
```diff
@@ -20,19 +20,62 @@ limitations under the License.
...
+    [[nodiscard]] auto end() const { return usedDecls.end(); }
+
+    [[nodiscard]] auto used() const { return Util::iterator_range(usedDecls); }
+
+    void clear() { usedDecls.clear(); }
+
+    void dbprint(std::ostream &cout) const override;
+
+    /// @returns @true if @p decl is used in the program.
+    bool isUsed(const IR::IDeclaration *decl) const { return usedDecls.count(decl) > 0; }
```
Comment (vlstill) [inline frontends/p4/unusedDeclarations.h:48]: I believe there is C++20 style `.contains` in the Abseil containers, it would be better to use that.

**Review by vlstill — COMMENTED** [review]
Thanks for all these changes!

I'm not sure about the changes to evaluator in frontend and inlining. From the comment in inlining, it seems like the pass itself does not need to re-run evaluator, that it is just to update its state for future use. But then the evaluator is not used again if it is not passed into Inlining. Also in the frontend, the evaluator "result" is not used.

As `Evaluator` is an inspector, presumably running it does not *change anything*, do we even need it in Frontend? Or is it because of errors it can produce? There seems to be just one...
https://github.com/p4lang/p4c/blob/7009f96b970054e4520af380a527b595d37e2300/frontends/p4/evaluator/evaluator.cpp#L133-L134

File: midend/flattenUnions.h [hunk 6]
```diff
@@ -165,14 +165,20 @@ class HandleValidityHeaderUnion : public Transform {
...
-    explicit RemoveUnusedHUDeclarations(P4::ReferenceMap *refMap) : refMap(refMap) {}
-    const IR::Node *preorder(IR::Type_HeaderUnion *type) {
-        if (!refMap->isUsed(getOriginal<IR::IDeclaration>())) {
-            return nullptr;
-        }
+    Visitor::profile_t init_apply(const IR::Node *node) override {
+        auto rv = Transform::init_apply(node);
+
+        CollectUsedDeclarations collector(used);
+        node->apply(collector);
```
[the remark this answers is not in the document]
Comment (asl) [reply midend/flattenUnions.h:175]: I'd rather make `RemoveUnusedHUDeclarations` a `PassManager`, similar to `RemoveAllUnusedDeclarations`

File: frontends/p4/unusedDeclarations.h [hunk 7]
```diff
@@ -20,19 +20,62 @@ limitations under the License.
...
+    absl::flat_hash_set<const IR::IDeclaration *, Util::Hash> usedDecls;
+
+ public:
+    bool setUsed(const IR::IDeclaration *decl) { return usedDecls.emplace(decl).second; }
+
+    [[nodiscard]] auto begin() const { return usedDecls.begin(); }
+
+    [[nodiscard]] auto end() const { return usedDecls.end(); }
+
+    [[nodiscard]] auto used() const { return Util::iterator_range(usedDecls); }
```
[the remark this answers is not in the document]
Comment (asl) [reply frontends/p4/unusedDeclarations.h:41]: Yeah, looks like a leftover.

**Comment by asl** [comment]
> As `Evaluator` is an inspector, presumably running it does not _change anything_, do we even need it in Frontend? Or is it because of errors it can produce? There seems to be just one...

Yeah, I was confused as well. And applied the common logic: if something looks incomplete, then probably there is a downstream code that uses / changes this. So, I tried to keep the existing logic as much as possible.

The Evaluator itself is a bit special:
 - It does its own traversal
 - While it is an Inspector, it creates new nodes. Though, it does not attach them anywhere, they are used to build a hierarchy of `CompileTimeValue` / `Block` nodes similar to CFG.

The inliner (`DiscoverInlining` specifically) does depend on the evaluator as it seems to traverse `Block`'s from the top level block, not the whole program. Given that `Inliner` runs in `PassRepeated` we'd need to update that CFG re-running evaluator.

I think we'd can just remove it from `Frontend`, yes.

**Comment by asl** [comment]
I think I addressed the majority of review comments. 

@fruffy I'm going to do `TypeChecking` changes in a separate PR.

**Comment by vlstill** [comment]
> The inliner (`DiscoverInlining` specifically) does depend on the evaluator as it seems to traverse `Block`'s from the top level block, not the whole program. Given that `Inliner` runs in `PassRepeated` we'd need to update that CFG re-running evaluator.

Oh, I missed that inliner runs as `PassRepeated`, sorry.

File: frontends/p4/unusedDeclarations.h [hunk 8]
```diff
@@ -144,15 +183,15 @@ class RemoveUnusedDeclarations : public Transform {
...
-    RemoveAllUnusedDeclarations(ReferenceMap *refMap, const RemoveUnusedPolicy &policy,
-                                bool warn = false)
-        : PassManager({new ResolveReferences(refMap),
-                       policy.getRemoveUnusedDeclarationsPass(refMap, warn)}) {
+    explicit RemoveAllUnusedDeclarations(const RemoveUnusedPolicy &policy, bool warn = false)
+        : PassManager({new CollectUsedDeclarations(used),
+                       policy.getRemoveUnusedDeclarationsPass(used, warn)}) {
         // Unused extern instances are not removed but may still trigger
         // warnings.  The @warned set keeps track of warnings emitted in
         // previous iterations to avoid emitting duplicate warnings.
```
Comment (vlstill) [inline frontends/p4/unusedDeclarations.h:194]: I think this comment belongs either to `RemoveUnusedDeclarations` class, or maybe to the `warned` member, but I don't see what it is for here, when `warned` does not appear here.

File: midend/flattenUnions.h [hunk 9]
```diff
@@ -165,18 +165,34 @@ class HandleValidityHeaderUnion : public Transform {
...
+
+ public:
+    RemoveAllUnusedHUDDeclarations()
+        : PassManager({
+              new CollectUsedDeclarations(used),
+              new RemoveUnusedHUDeclarations(used),
+          }) {
+        // Unused extern instances are not removed but may still trigger
+        // warnings.  The @warned set keeps track of warnings emitted in
+        // previous iterations to avoid emitting duplicate warnings.
```
Comment (vlstill) [inline midend/flattenUnions.h:190]: same as above, I think this comment is misplaced.

```suggestion

**Review by vlstill — APPROVED** [review]

File: frontends/p4/unusedDeclarations.h (hunk 8 above)
Comment (asl) [reply frontends/p4/unusedDeclarations.h:194]: Yeah, likely it left after some earlier refactoring. I'll fix it while here.

**Comment by asl** [comment]
Looks like both bazel and fedora are unhealthy. @grg @fruffy do you know what's going on?

**Comment by fruffy** [comment]
> Looks like both bazel and fedora are unhealthy. @grg @fruffy do you know what's going on?

It's possible that one of the new includes is breaking the Bazel build related to alloc_trace. For Fedora, unclear how that error happens. Hope it is spurious.

**Review by ChrisDodd — APPROVED** [review]

**Comment by fruffy** [comment]
@smolkaj Any idea whether the Bazel failures are spurious or require a change?

**Comment by asl** [comment]
@fruffy @smolkaj Bazel seems to be recovered. Only Fedora left...

**Pull request merged**